azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +101,232 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
         else:
             logger.error(msg)
         if send_event:
-            add_event(op=op, message=msg, is_success=is_success)
+            add_event(op=op, message=msg, is_success=is_success, log_event=False)
+
+    @staticmethod
+    def __get_policy_enforcement_enabled():
+        """
+        Policy will be enabled if (1) policy file exists at the expected location and (2) the conf flag "Debug.EnableExtensionPolicy" is true.
+        """
+        return conf.get_extension_policy_enabled() and os.path.exists(conf.get_policy_file_path())
+
+    @property
+    def policy_enforcement_enabled(self):
+        return self._policy_enforcement_enabled
+
+    @staticmethod
+    def __get_policy():
+        """
+        Read customer-provided policy JSON file, load and return as a dict.
+        Policy file is expected to be at conf.get_policy_file_path(). Note that this method should only be called
+        after verifying that the file exists (currently done in __init__).
+        """
+        with open(conf.get_policy_file_path(), 'r') as f:
+            _PolicyEngine._log_policy_event(
+                "Policy enforcement is enabled. Enforcing policy using policy file found at '{0}'.".format(
+                    conf.get_policy_file_path()))
+            try:
+                # json.load will raise error if file is not in valid json format (including empty file).
+                custom_policy = json.load(f)
+            except ValueError as ex:
+                msg = "policy file does not conform to valid json syntax"
+                raise InvalidPolicyError(msg=msg, inner=ex)
+            except Exception as ex:
+                msg = "unable to load policy file"
+                raise InvalidPolicyError(msg=msg, inner=ex)
+
+            return custom_policy
+
+    def __parse_policy(self, custom_policy):
```

Comment:
It's confusing that __parse_atribute has a return value and the rest of the parse functions don't. can we make them consistent?

Reply:
I've removed __parse_attribute, type validation is now done in __validate_schema